storage/mysql/storage_test.go

```diff
@@ -552,6 +552,82 @@ func TestDequeueLeavesGuardInterval(t *testing.T) {
 	}
 }
 
+func TestDequeueLeavesTimeOrdering(t *testing.T) {
+	// Queue two small batches of leaves at different timestamps. Do two separate dequeue
+	// transactions and make sure the returned leaves are respecting the time ordering of the
+	// queue.
+
+	logID := createLogID("TestDequeueLeavesTimeOrdering")
+	db := prepareTestLogDB(logID, t)
+	defer db.Close()
+	s := prepareTestLogStorage(logID, t)
+	batchSize := 2
+
+	{
+		tx := beginLogTx(s, t)
+		defer failIfTXStillOpen(t, "TestDequeueLeavesTimeOrdering", tx)
+
+		leaves := createTestLeaves(int64(batchSize), 0)
+		leaves2 := createTestLeaves(int64(batchSize), int64(batchSize))
+
+		if err := tx.QueueLeaves(leaves, fakeQueueTime); err != nil {
+			t.Fatalf("Failed to queue leaves: %v", err)
+		}
+
+		// These are one second earlier so should be dequeued first
+		if err := tx.QueueLeaves(leaves2, fakeQueueTime.Add(-time.Second)); err != nil {
+			t.Fatalf("Failed to queue leaves2: %v", err)
+		}
+
+		commit(tx, t)
+	}
+
+	{
+		// Now try to dequeue two leaves and we should get the second batch
+		tx2 := beginLogTx(s, t)
+		defer failIfTXStillOpen(t, "TestDequeueLeavesTimeOrdering", tx2)
+		dequeue1, err := tx2.DequeueLeaves(batchSize, fakeQueueTime)
+
+		if err != nil {
+			t.Fatalf("Failed to dequeue leaves (1st): %v", err)
+		}
+
+		if got, want := len(dequeue1), batchSize; got != want {
+			t.Fatalf("Dequeue count mismatch (1st) got: %d, want: %d", got, want)
+		}
+
+		ensureAllLeafHashesDistinct(dequeue1, t)
+
+		// Ensure this is the second batch queued by comparing leaf data.
+		if !leafInRange(dequeue1[0], batchSize, batchSize + batchSize - 1) || !leafInRange(dequeue1[1], batchSize, batchSize + batchSize - 1) {
+			t.Fatalf("Got leaf from wrong batch (1st dequeue): (%s %s) %v", string(dequeue1[0].LeafValue), string(dequeue1[1].LeafValue), dequeue1)
```

Comment:
I'll make it just print the mismatched strings. The rest of the leaf contents don't really matter in this context.